Move allgather_coalesced implementation from Python to C++

agolynski opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/29059

Resubmit of reverted PR #28857.

Differential Revision: D18277097

zappa- commented:
https://github.com/pytorch/pytorch/pull/29059 caused a broken build due to unimplemented function in MPI backend. Fixed here.

pietern reviewed (approved):
I should have checked CI before approving #28857.

It's all green now, so it should be good to go.

agolynski committed:
Moving python allgather_coalesced impl from Py to C. (#29059)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/29059
This is a resubmit of reverted diff D18209289 ( PR #28857 ).

Test Plan:
buck test caffe2/test:c10d
buck test caffe2/test:distributed_gloo

Reviewed By: pietern

Differential Revision: D18277097

fbshipit-source-id: 3e16c4c5f71e5c051ffef280e021bd253caf127c

gchanan commented on test/test_c10d.py:
```diff
@@ -1309,31 +1309,37 @@ def test_allgather_stress_cuda(self):
         store = c10d.FileStore(self.file_name, self.world_size)
         pg = c10d.ProcessGroupGloo(store, self.rank, self.world_size, self.opts())
-        dummy_input = [torch.Tensor([1])]
+        dummy_input = [torch.zeros([1], dtype=torch.float32)]
```
sorry -- why are these not translated exactly?

`torch.Tensor([1])` is `torch.ones([1])`, not zeros, right?

also same with the line below, why did that change from -1 to 0?

zappa- replied:
This only tests error handling, so they underlying values here should not be important (all_gather_coalesced never copies anything in this function). I am happy to change it back if you prefer

gchanan commented on torch/lib/c10d/Utils.hpp:
```diff
@@ -203,6 +203,20 @@ inline void assertCPU(
 
+inline void assertSameDevice(
+    std::function<void(const std::string&)> fn,
+    const at::ArrayRef<at::Tensor>& tensors) {
```
don't we have a TensorList for this?  (Also I wouldn't expect const reference to it, it's trivial to copy).

zappa- replied:
ah, a good point. I was trying to be consistent with the other functions in this module mostly use const reference to ArrayRef, and not TensorList (TensorList = ArrayRef<Tensor>).
Actually, I just need to verify tensors in a vector, so I might just accept a const ref to a vector.
Would you prefer that?
